src/advanced/utils/validators.ts

```diff
@@ -0,0 +1,113 @@
+// 검증 유틸리티 함수들
+// 구현할 함수:
+// - isValidCouponCode(code: string): boolean - 쿠폰 코드 형식 검증 (4-12자 영문 대문자와 숫자)
+// - isValidStock(stock: number): boolean - 재고 수량 검증 (0 이상)
+// - isValidPrice(price: number): boolean - 가격 검증 (양수)
+// - extractNumbers(value: string): string - 문자열에서 숫자만 추출
+
+export const isValidStock = (stock: string) => {
+  if (stock === "") {
+    return {
+      isValid: true,
+      stock: 0,
+      message: "",
+    };
+  } else if (parseInt(stock) < 0) {
+    return {
+      isValid: false,
+      stock: 0,
+      message: "재고는 0보다 커야 합니다",
+    };
+  } else if (parseInt(stock) > 9999) {
+    return {
+      isValid: false,
+      stock: 9999,
+      message: "재고는 9999개를 초과할 수 없습니다",
+    };
+  }
+  return {
+    isValid: true,
+    stock: parseInt(stock),
+    message: "",
+  };
+};
+
+export const isValidPrice = (price: string) => {
+  if (price === "") {
+    return {
+      isValid: true,
+      price: 0,
+      message: "",
+    };
+  } else if (parseInt(price) < 0) {
+    return {
+      isValid: false,
+      price: 0,
+      message: "가격은 0보다 커야 합니다",
+    };
+  }
+
+  return {
+    isValid: true,
+    price: parseInt(price),
+    message: "",
+  };
+};
+
+export const isValidDiscount = (discountType: string, value: number) => {
+  if (discountType === "percentage") {
```

Comment:
“percentage” 같은 하드코딩된 문자열은 오타나 변경 시 추적이 어려울 수 있어서 상수(enum)나 상수 객체로 관리하면 타입 안정성을 높일 수 있어요! 이렇게 하면 자동완성도 지원돼서 아주 편리합니다.

```suggestion
// constants/discount.ts

// 객체 방식
export const DISCOUNT_TYPE = {
  PERCENTAGE: 'percentage',
  AMOUNT: 'amount',
} as const;

export type DiscountType = typeof DISCOUNT_TYPE[keyof typeof DISCOUNT_TYPE];

// 사용 예시
if (discountType === DISCOUNT_TYPE.PERCENTAGE) {
  // ...
}

// enum 방식
export enum DiscountTypeEnum {
  PERCENTAGE = 'percentage',
  AMOUNT = 'amount',
}

// 사용 예시
if (discountType === DiscountTypeEnum.PERCENTAGE) {
  // ...
}
```